torch/distributed/_checkpoint/state_dict_saver.py

```diff
@@ -0,0 +1,214 @@
+    )
+    metadata.state_dict_metadata[fqn] = etmd
+
+
+def _prepare(state_dict: Dict[str, Any]) -> Tuple[Metadata, Dict[str, int], List[ReadWriteRequest]]:
+    """
+    Uses the state_dict to build three things.
+
+    metadata: Metadata
+        The metatdata discribing the tensor / sharded tensor.
+        And it is storage meta data. See "../metadata.py" for detail
+
+    size_for_storage_handles: Dict[str, int]
+        Key is the handle name, value is the handle's size
+        It can used to pre allocate the storage for parallel and non sequential writes.
+
+    write_requests: List[ReadWriteRequest]
+        List of write requests that should br perfromed by the writer.
+
+
+    Subclasses can optionally overwrite the implementation here,
+    if the default does not meet its requirement.
+    """
+    metadata = Metadata(state_dict_metadata={})
+    storage_handles = {}
+    write_requests = []
+
+    for fqn, tensor in state_dict.items():
+        if isinstance(tensor, Tensor):
+            # The assumption is that non ShardedTensors are full replicated across all ranks
+            # So we just need one from Rank 0.
+            # If that's not the case, we will update later.
+            if dist.is_initialized() and dist.get_rank() != 0:
+                pass
+            else:
+                _populate_inplace_with_a_tensor(fqn, tensor, metadata, storage_handles, write_requests)
+        elif isinstance(tensor, ShardedTensor):
+            _populate_inplace_with_a_sharded_tensor(fqn, tensor, metadata, storage_handles, write_requests)
+        else:
+            raise RuntimeError("The input need to be either Tensor or ShardedTensor")
+
+    return (metadata, storage_handles, write_requests)
+
+
+# These two public functions defined the default behavior to save a state_dict
+# Note this is a WIP, the state_dict save with different version of the code might not be
+# compatible.
+#
+# This code defines/determines these schema
+# 1. The metadata that discribe the state_dict.
+# 2. How we map each tensor/sharded_tensor to storage handles.
+def save_state_dict(
+    state_dict: Dict[str, Any],
+    storage_writer: StorageWriter,
+    metadata_prepare_fn : Callable = _prepare,
+) -> None:
+    """
+    The same as save_state_dict_async, but blocking until everything is finished.
+
+    Args:
+        state_dict (Dict[str, Any]) : A state_dict
+        storage_writer (StorageWriter): An instance of storage writer that
+            performance the writes.
+        metadata_prepare_fn (Callable): The function what creates
+            the metadata, write request and others.
+    """
+    futures = save_state_dict_async(
+        state_dict=state_dict,
+        storage_writer=storage_writer,
+        metadata_prepare_fn=metadata_prepare_fn,
+    )
+
+    torch.futures.wait_all(futures)
+
+def save_state_dict_async(
+    state_dict: Dict[str, Any],
+    storage_writer: StorageWriter,
+    metadata_prepare_fn: Callable = _prepare,
```

Comment:
Should we add type annotation for like `Callable[[Dict[str,Any]], Tuple(Any,Any,Any)]` ?